nexus/authz-macros/src/lib.rs

```diff
@@ -278,7 +278,7 @@ fn do_authz_resource(
 
     Ok(quote! {
         #[doc = #doc_struct]
-        #[derive(Clone, Debug)]
+        #[derive(Clone, Debug, Serialize, Deserialize)]
```

Comment:
I think we talked about this offline or I at least raised the question.  This makes these internal structures a committed interface because they'll be stored into the database.  Is that not a problem?

Reply:
I thought this was the topic of RFD 289: [Steno Upgrade](https://rfd.shared.oxide.computer/rfd/289).

> [Determinations](https://rfd.shared.oxide.computer/rfd/289#_determinations)
> For a given saga, execution is only ever carried out by one version of Nexus. Failover may still cause execution to move between different instances of Nexus.
>
> There are two pieces to this: draining sagas during upgrade and having Nexus recover only sagas that were created with the same version as itself.

If we *don't* do this, then Nexus will need to redo lookup of relevant resources repeatedly within each saga step. In addition to the cost of looking up / recording saga states, this seems like it amplifies the total number of database operations needed to complete the "real work" within a saga.

Reply:
I also view this as a major tension between:

- `authz` objects are becoming more-and-more a required input to datastore operations...
- ... and many saga operations want to issue exactly one datastore operation.

So this leaves us with a few options:
1. **No serialization, authz is still required for datastore ops**: This means we need to repeatedly lookup each resource we're acting upon, and inflates the total number of requests to the db. This also means that a permission change mid-operation could cause some odd failure modes.
2. **Do authorization checks earlier**. Basically, rather than passing an `authz` object to datastore for evaluation, pass around an object that implies "the caller is *already authorized to perform the operation*". This new object would need to be serializable and provided as input to datastore functions, but the `authz::Resource` objects could remain outside the saga log.
3. **Serialize the `authz::Resource` objects, use them in the datastore operations as they exist today**. This is basically what this PR does, by adding the derives.

Reply:
> I thought this was the topic of RFD 289

Yeah, good call.  I think as long as we only ever serialize these into saga state (and not elsewhere in the database, for example, or over HTTP interfaces), I think you're right that this is fine.